src/lib_db.ml

```diff
@@ -165,34 +178,40 @@ module Scope = struct
     required_in_jbuild scope ~jbuild_dir:dir
 
   (* Fold the transitive closure, not necessarily in topological order *)
-  let fold_transitive_closure scope ~deep_traverse_externals lib_deps ~init ~f =
+  let fold_transitive_closure (scope : t With_required_by.t)
+        ~deep_traverse_externals lib_deps ~init ~f =
     let seen = ref String_set.empty in
     let rec loop scope acc lib_dep =
-      match interpret_lib_dep scope lib_dep with
-      | Inr fail -> fail.fail ()
-      | Inl libs -> List.fold_left libs ~init:acc ~f:process
+      interpret_lib_dep_exn scope lib_dep
+      |> List.fold_left ~init:acc ~f:process
     and process acc (lib : Lib.t) =
       let unique_id = Lib.unique_id lib in
       if String_set.mem unique_id !seen then
         acc
       else begin
         seen := String_set.add unique_id !seen;
         let acc = f lib acc in
-        match lib with
-        | Internal (dir, lib) ->
-          let scope = find_scope' scope.With_required_by.data.lib_db ~dir in
-          List.fold_left lib.buildable.libraries ~init:acc ~f:(loop scope)
-        | External pkg ->
-          if deep_traverse_externals then
-            List.fold_left (FP.requires pkg) ~init:acc ~f:(fun acc pkg ->
-              process acc (Lib.external_ pkg))
-          else begin
-            seen :=
-              String_set.union !seen
-                (String_set.of_list
-                   (List.map (FP.requires pkg) ~f:FP.name));
-            acc
-          end
+        let requires = Lib.requires lib in
+        let scope =
+          match Lib.scope lib with
+          | `External ->
+            { With_required_by.
+              data = external_scope scope.data.lib_db
+            ; required_by = scope.required_by
+            }
```

Comment:
We can do a little better by passing the old scope through loop/process to use the most recent `required_by` in the scope stack.